Task7/vertex-ai-qwikstart/utils/data_download.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import logging
+import shutil
+import wget
+import argparse
+import pandas as pd
+from google.cloud import storage
+from google.cloud import bigquery
+from google.cloud.exceptions import NotFound, Conflict
+
+from dataset_schema import table_schema
+from dataset_clean import dataset_clean_query
+from dataset_ml import dataset_ml_query
+
+LOCAL_PATH ="./data"
+FILENAME = "online_retail"
+
+
+def download_url2gcs(args):
+    """
+    args:
+    """
+    
+    #set GCS client.
+    client = storage.Client()
+    
+    # Retrieve GCS bucket.
+    bucket = client.get_bucket(args.GCS_BUCKET)
+    blob = bucket.blob("data/online_retail.csv")
+
+    #See if file already exists.
+    if blob.exists() == False:
+        try:
+            os.mkdir(LOCAL_PATH)
+            logging.info('Downloading xlsx file...')            
+            local_xlsx = wget.download(args.URL, out=f"{LOCAL_PATH}/{FILENAME}.xlsx")
+            logging.info('Converting xlsx -> csv...')            
+            df = pd.read_excel(local_xlsx)            
+            df.to_csv(f"{LOCAL_PATH}/{FILENAME}.csv", index=False)
+            logging.info('Uploading local csv file to GCS...')
+            blob.upload_from_filename(f"{LOCAL_PATH}/{FILENAME}.csv")
+            logging.info('Copied local csv file to GCS.')
+            # Delete all contents of a directory using shutil.rmtree() and  handle exceptions.
+            try:
+                shutil.rmtree(LOCAL_PATH)
+                logging.info('Cleaning up local tmp data directory...')
+            except:
+                logging.error('Error while deleting local tmp data directory.')
+
+        #print error if file doesn't exist.
+        except BaseException as error:
+            logging.error('An exception occurred: {}'.format(error))
+            
+    #print error if file already exists in GCS.
+    else:
+        logging.warning('File already exists in GCS.')
```

Comment:
```suggestion
    if blob.exists():
        logging.warning('File already exists in GCS.')
        # TODO fix indent
      try:
          os.mkdir(LOCAL_PATH)
          logging.info('Downloading xlsx file...')            
          local_xlsx = wget.download(args.URL, out=f"{LOCAL_PATH}/{FILENAME}.xlsx")
          logging.info('Converting xlsx -> csv...')            
          df = pd.read_excel(local_xlsx)            
          df.to_csv(f"{LOCAL_PATH}/{FILENAME}.csv", index=False)
          logging.info('Uploading local csv file to GCS...')
          blob.upload_from_filename(f"{LOCAL_PATH}/{FILENAME}.csv")
          logging.info('Copied local csv file to GCS.')
          # Delete all contents of a directory using shutil.rmtree() and  handle exceptions.
          try:
              shutil.rmtree(LOCAL_PATH)
              logging.info('Cleaning up local tmp data directory...')
          except:
              logging.error('Error while deleting local tmp data directory.')

      #print error if file doesn't exist.
      except BaseException as error:
          logging.error('An exception occurred: {}'.format(error))
            

```

Reply:
I am not sure I get this logic
- If the bucket does not exist yet in the Cloud Storage, it will fail
- If the bucket exists, then we will overwrite it with

Reply:
I am not sure I understand this logic, so if the bucket doesn't yet exist, we just skip the whole `if` statement and no new bucket is created

Reply:
I'm missing an indent I see. I was just trying to move the short else block with the warning up to avoid these large if-else blocks. Should the try block be run always or only if the file does not exist?

Reply:
The try block should only be run if the file does not exist. (If the file does not yet exist in GCS, we download it from a given URL and then upload it)